app/models/episode.rb

```diff
@@ -133,28 +129,75 @@ def author=(a)
     self.author_email = author["email"]
   end
 
-  def enclosure
-    enclosures.complete.first
+  def ready_image
+    images.complete_or_replaced.first
   end
 
   def image
-    images.complete.first
-  end
-
-  # API updates for image=
-  def image_file
     images.first
   end
 
-  def image_file=(file)
+  def image=(file)
     img = EpisodeImage.build(file)
-    if img && img.original_url != image_file.try(:original_url)
-      images << img
-    elsif !img
-      images.destroy_all
+
+    if !img
+      images.each(&:mark_for_destruction)
+    elsif img&.replace?(image)
+      images.build(img.attributes.compact)
+    else
+      img.update_image(image)
+    end
+  end
+
+  def ready_enclosure
+    enclosures.complete_or_replaced.first
+  end
+
+  def enclosure
+    enclosures.first
+  end
+
+  def enclosure=(file)
+    enc = Enclosure.build(file)
+
+    if !enc
+      enclosures.each(&:mark_for_destruction)
+    elsif enc&.replace?(enclosure)
+      enclosures.build(enc.attributes.compact)
+    else
+      enc.update_resource(enclosure)
     end
   end
 
+  def ready_contents
+    contents.complete_or_replaced.group_by(&:position).values.map(&:first)
```

Comment:
Here's another pattern I'm following:

The public RSS (or the API JSON that DTR uses) should always return "ready" files.  Which may be the latest/greatest `contents` here.  Or it might be an older file we just replaced.

Reply:
I guess my thinking was... we generally want our APIs and HTML pages to show users the latest state.  What files are in this episode, are they processing, etc.  And also, it made more sense to have the default `has_many` be a normal (doesn't include deleted) relation?  I was trying to get away from the more magical multi-relations we had before, as they were causing issues.

And the exception is for the Feed RSS and DTR - where we want to gatekeep only completed files.

I guess it depends on perspective?